packages/ember-template-compiler/lib/plugins/assert-modifiers-not-in-components.ts

```diff
@@ -0,0 +1,54 @@
+import { assert } from '@ember/debug';
+import { AST, ASTPlugin, ASTPluginEnvironment } from '@glimmer/syntax';
+import calculateLocationDisplay from '../system/calculate-location-display';
+
+export default function assertModifiersNotInComponents(env: ASTPluginEnvironment): ASTPlugin {
+  let { moduleName } = env.meta;
+  let scopes: string[][] = [];
+
+  function isComponentInvocation(node: AST.ElementNode) {
+    return (
+      node.tag[0] === '@' ||
+      node.tag[0].toUpperCase() === node.tag[0] ||
+      node.tag.indexOf('.') > -1 ||
+      scopes.some(params => params.some(p => p === node.tag))
+    );
+  }
+
+  return {
+    name: 'assert-modifiers-not-in-components',
+
+    visitor: {
+      Program: {
+        enter(node: AST.Program) {
+          scopes.push(node.blockParams);
+        },
+
+        exit() {
+          scopes.pop();
+        },
+      },
+      ElementNode: {
+        keys: {
+          children: {
+            enter(node: AST.ElementNode) {
+              scopes.push(node.blockParams);
+            },
+
+            exit() {
+              scopes.pop();
+            },
+          },
+        },
+        enter(node: AST.ElementNode) {
+          if (node.modifiers.length > 0 && isComponentInvocation(node)) {
+            assert(`Modifiers cannot be applied to components. ${calculateLocationDisplay(
```

Comment:
nit: since we are about to roll this out/tech this to people, let’s avoid the “cannot be applied” wording, and go for something like “Passing modifiers to components require the xyz canary feature, which has not been stabilized yet. See RFC xyz for details.”